### PR TITLE
Integrate emoji with wallet

### PR DESCRIPTION
**Pull request by xitronix**

### CODE REVIEW

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 1]
```diff
@@ -3,19 +3,23 @@ import {EmojiPlaceholders} from './EmojiPlaceholders';
 import {EmojiPanelWithFakes} from './EmojiPanelWithFakes';
 import {isValidCode} from '@universal-login/commons';
 import UniversalLoginSDK from '@universal-login/sdk';
+import {transactionDetails} from '../../core/constants/TransactionDetails';
 
 interface EmojiFormProps {
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Let's simplify Props:
```ts
interface EmojiFormProps {
  publicKey: string;
  onCancel: Function
  onSuccess: Function
}
```

Comment (marekkirejczyk) [reply universal-login-react/src/ui/Notifications/EmojiForm.tsx]: `onCancel` -> sdk.denyRequest(contractAddress, publicKey, privateKey)
`onSuccess` -> sdk.addKey(contractAddress, publicKey, privateKey, transactionDetails);
`securityCodeWithFakes` -> calculate from publicKey

File: universal-login-wallet/src/ui/react/ConnectAccount/HandleConnection.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import React, {useState, useEffect} from 'react';
+import {EmojiForm} from '@universal-login/react';
+import {Notification} from '@universal-login/commons';
+import vault1x from './../../assets/illustrations/vault.png';
+import vault2x from './../../assets/illustrations/vault@2x.png';
+import {useServices} from '../../hooks';
+import {Link} from 'react-router-dom';
+
+export const HandleConnection = () => {
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/ConnectAccount/HandleConnection.tsx]: `ConnectionNotification`?

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 3]
```diff
@@ -3,19 +3,23 @@ import {EmojiPlaceholders} from './EmojiPlaceholders';
...
 
-export const EmojiForm = ({sdk, publicKey, securityCodeWithFakes}: EmojiFormProps) => {
+export const EmojiForm = ({sdk, publicKey, privateKey, contractAddress, securityCodeWithFakes}: EmojiFormProps) => {
   const [enteredCode, setEnteredCode] = useState([] as number[]);
   const [status, setStatus] = useState('');
 
   const confirmWithCodeCheck = (publicKey: string) => {
     if (isValidCode(enteredCode, publicKey)) {
+      sdk.addKey(contractAddress, publicKey, privateKey, transactionDetails);
       setStatus('OK: security code confirmed');
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: Set status to empty.

File: universal-login-react/src/ui/Notifications/EmojiForm.tsx [hunk 4]
```diff
@@ -3,19 +3,23 @@ import {EmojiPlaceholders} from './EmojiPlaceholders';
...
     } else {
       setStatus('FAIL: Wrong security code. Try again or deny request.');
```
Comment (marekkirejczyk) [inline universal-login-react/src/ui/Notifications/EmojiForm.tsx]: setStatus to `Invalid code. Try again.`

File: universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx [hunk 5]
```diff
@@ -1,25 +1,35 @@
...
-            <p className="box-text connect-emoji-text">Thanks, now check another device controling this account and enter the emojis in this order:</p>
-            <EmojiPanel code={generateCode(TEST_ACCOUNT_ADDRESS)}/>
-            <button className="button-secondary connect-emoji-btn">Deny</button>
+interface ConnectWithEmojiProps {
+  name: string;
+}
+
+export const ConnectWithEmoji = ({name}: ConnectWithEmojiProps) => {
+  const {connectToWallet} = useServices();
+  const [connectValues] = useAsync(async () => connectToWallet(name, () =>  history.push('/')), []);
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: extract:
`const navigateToMainScreen = () => history.push('/')` and pass as ConnectWithEmoji prop `onSuccess`.

File: universal-login-wallet/src/core/services/ConnectToWallet.ts [hunk 6]
```diff
@@ -4,11 +4,18 @@ import {Wallet} from 'ethers';
...
   walletService.connect({privateKey, contractAddress, name});
   const filter = {contractAddress, key: publicKey};
-  const subscription = sdk.subscribe('KeyAdded', filter, callback);
-  return () => subscription.remove();
+  const unsubscribe = () => subscription.remove();
+  const subscription = sdk.subscribe('KeyAdded', filter, () => {
+    unsubscribe();
+    callback();
+  });
+  return {
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/core/services/ConnectToWallet.ts]: Add type, so there is no need to handle null.

File: universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx (hunk 5 above)
Comment (marekkirejczyk) [reply universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: `const [{securityCode}, error] = `

Add error handling

File: universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx [hunk 7]
```diff
@@ -1,25 +1,35 @@
...
+      <div className="box-wrapper">
+        <div className="box">
+          <div className="box-header">
+            <h1 className="box-title">Confirmation</h1>
+          </div>
+          <div className="box-content connect-emoji-content">
+            <div className="connect-emoji-section">
+              <img src={vault1x} srcSet={vault2x} alt="avatar" className="connect-emoji-img" />
+              <p className="box-text connect-emoji-text">Thanks, now check another device controling this account and enter the emojis in this order:</p>
+              {connectValues ? <EmojiPanel code={connectValues!.securityCode} /> : 'Loading'}
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: if (error) {
  error
} else if (securityCode) {
  <EmojiPanel code={securityCode} />
} else {
  'Loading...'
}

File: universal-login-wallet/test/integration/pages/Notifications.ts [hunk 8]
```diff
@@ -5,20 +5,28 @@ export default class NotificationsPage {
...
+    this.wrapper.find('#reject').simulate('click');
     await this.waitForNotificationDisappear();
   }
 
+  async inputSecurityCode(securityCode: number[]) {
+    await waitForUI(this.wrapper, () => this.wrapper.text().includes('Security code'));
+    for (const number of securityCode) {
+      const button = this.wrapper.find(`#btn-${number}`).first();
+      button.simulate('click');
+      await waitForUI(this.wrapper, () => this.wrapper.find(`img[alt="${number}"]`).length === 2);
```
Comment (marekkirejczyk) [inline universal-login-wallet/test/integration/pages/Notifications.ts]: What if we have repetitions? Shouldn't we first check the number of occurrences that checked it was increased?

File: universal-login-wallet/src/ui/react/ConnectAccount/HandleConnection.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            <div className="connect-emoji-section">
+              <img src={vault1x} srcSet={vault2x} alt="avatar" className="connect-emoji-img" />
+              <p className="box-text connect-emoji-text">Thanks, now check another device controling this account and enter the emojis in this order:</p>
+              {notifications.length > 0 && <EmojiForm
+                securityCodeWithFakes={notifications[0].securityCodeWithFakes!}
+                publicKey={notifications[0].key}
+                sdk={sdk}
+                contractAddress={contractAddress}
+                privateKey={privateKey}
+              />
```
Comment (marekkirejczyk) [inline universal-login-wallet/src/ui/react/ConnectAccount/HandleConnection.tsx]: If there is no connection, we should show 'No requests to connect from other applications'.

File: universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx (hunk 5 above)
Comment (marekkirejczyk) [reply universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: `connectValues.unsubscribe` is never used, shouldn't it be used?  @JustynaBroniszewska

Comment (JustynaBroniszewska) [reply universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: Yes, we should use `connectValues .unsubscribe` on the `deny` button click.

Comment (xitronix) [reply universal-login-wallet/src/ui/react/ConnectAccount/ConnectWithEmoji.tsx]: 1. We cannot extract () => history.push('/'). When we do it we breaking the rule 'Don’t call Hooks from regular JavaScript functions'
2. We need to have the possibility to really deny the request by sdk. Not just unsubscribe. Next PR?

File: universal-login-wallet/src/core/services/ConnectToWallet.ts (hunk 6 above)
Comment (xitronix) [reply universal-login-wallet/src/core/services/ConnectToWallet.ts]: Type added

**Pull request merged**